### ISSUE TITLE
4425 - Greater Than - arbitrary sized numbers

### ISSUE THREAD

**Issue by Alexsey** [type-challenges/type-challenges#21721]
```ts
// compares the first digits of T and U. If they are the same - compare rest of T and rest of U
type GreaterThan<T extends number | string, U extends number | string> =
  `${T}` extends `${infer TF}${infer TR}`
    ? `${U}` extends `${infer UF}${infer UR}`
      ? TF extends UF
        ? GreaterThan<TR, UR>
        : `0123456789` extends `${string}${TF}${string}${UF}${string}`
          ? false
          : true
      : true
    : false
```
Additional test cases, that don't work for solutions based on huge arrays building, but work with this solution:
```ts
Expect<Equal<GreaterThan<1234567891011, 1234567891010>, true>>,
// for numbers greater than 9007199254740991 (Number.MAX_SAFE_INTEGER) use strings
Expect<Equal<GreaterThan<'12345678910111213141516', '12345678910111213141515'>, true>>,
```

**Comment by Yueleng** [comment]
this is the wrong answer,

Try this in your test cases `Expect<Equal<GreaterThan<100, 99>, true>>`

**Comment by Yueleng** [comment]
I improved your answer as follows: 

```
// compares the first digits of T and U. If they are the same - compare rest of T and rest of U
// prettier-ignore
type GreaterThanSameDigitCount<
  T extends number | string,
  U extends number | string
> = `${T}` extends `${infer TF}${infer TR}`
  ? `${U}` extends `${infer UF}${infer UR}`
    ? TF extends UF
      ? GreaterThanSameDigitCount<TR, UR>
      : "0123456789" extends `${string}${TF}${string}${UF}${string}`
        ? false
        : true
    : true
  : false;

type DigitsToArr<S extends string> = S extends `${string}${infer R}`
  ? [0, ...DigitsToArr<R>]
  : [];

type ArrLenCompare<
  T extends any[],
  U extends any[]
> = "0123456789" extends `${string}${T["length"]}${string}${U["length"]}${string}`
  ? -1
  : "0123456789" extends `${string}${U["length"]}${string}${T["length"]}${string}`
  ? 1
  : 0;

type GreaterThan<T extends number, U extends number> = ArrLenCompare<
  DigitsToArr<`${T}`>,
  DigitsToArr<`${U}`>
> extends 0
  ? GreaterThanSameDigitCount<T, U>
  : ArrLenCompare<DigitsToArr<`${T}`>, DigitsToArr<`${U}`>> extends 1
  ? true
  : false;

```

**Comment by Alexsey** [comment]
Thank you, you are right - there was a bug! I've updated the solution to handle it and I'll create a PR to add your test case to the problem

**Comment by wkz2003** [comment]
[Res, TF & UF] extends [false, number] // Res == false && TF != UF
In this place,I think TF nad UF will be string type,why you wirte number instead? Am I wrong?

**Comment by Alexsey** [comment]
You are correct - it's a logical bug - it should have been `extends [false, never]`, but because both values are strings, `number` work like `never` :) It was very confusing - I've fixed it. Thank you!

**Comment by xlboy** [comment]
> I improved your answer as follows:
> 
> ```
> // compares the first digits of T and U. If they are the same - compare rest of T and rest of U
> // prettier-ignore
> type GreaterThanSameDigitCount<
>   T extends number | string,
>   U extends number | string
> > = `${T}` extends `${infer TF}${infer TR}`
>   ? `${U}` extends `${infer UF}${infer UR}`
>     ? TF extends UF
>       ? GreaterThanSameDigitCount<TR, UR>
>       : "0123456789" extends `${string}${TF}${string}${UF}${string}`
>         ? false
>         : true
>     : true
>   : false;
> 
> type DigitsToArr<S extends string> = S extends `${string}${infer R}`
>   ? [0, ...DigitsToArr<R>]
>   : [];
> 
> type ArrLenCompare<
>   T extends any[],
>   U extends any[]
> > = "0123456789" extends `${string}${T["length"]}${string}${U["length"]}${string}`
>   ? -1
>   : "0123456789" extends `${string}${U["length"]}${string}${T["length"]}${string}`
>   ? 1
>   : 0;
> 
> type GreaterThan<T extends number, U extends number> = ArrLenCompare<
>   DigitsToArr<`${T}`>,
>   DigitsToArr<`${U}`>
> > extends 0
>   ? GreaterThanSameDigitCount<T, U>
>   : ArrLenCompare<DigitsToArr<`${T}`>, DigitsToArr<`${U}`>> extends 1
>   ? true
>   : false;
> ```

That's a great answer. 
In addition, I translated it to make it easier to read. 😁
https://type-zen-playground.vercel.app/?example=type-challenges-medium-52_greater-than

<img width="778" alt="image" src="https://user-images.githubusercontent.com/63690944/227148106-f6264348-0a57-41ee-8b78-3ad710038e04.png">

**Comment by psmolak** [comment]
What a clever way to check `a < b`, nice @Alexsey 
```ts
'0123456789' extends `${string}${TF}${string}${UF}${string}
```

**Comment by Alexsey** [comment]
@psmolak thank you, I was also pretty impressed when I came up with this solution :)

**Comment by likepeach789** [comment]
What does this line mean? Could someone please help? '0123456789' extends `${string}${TF}${string}${UF}${string}` ? false : true

**Comment by Alexsey** [comment]
@likepeach789 sure!

this generic resolves to `true` when `TF` is a digit that is less than `UF`. E.g.
* `TF` is `3` and `UF` is `5` -> `true`
* `TF` is `3` and `UF` is `3` -> `false`
* `TF` is `3` and `UF` is `1` -> `false`

Let's check how it works by example:
* when `TF` is `3` and `UF` is `5` it will check if `...3...5...` matches `0123456789`. It is, so `true`
* when `TF` is `3` and `UF` is `3` it will check if `...3...3...` matches `0123456789`. It is **not** so `false`
* when `TF` is `3` and `UF` is `1` it will check if `...3...1...` matches `0123456789`. It is **not** so `false`

I hope the explanation is sufficient :) Please ask if it not

**Comment by likepeach789** [comment]
> @likepeach789 sure!
> 
> this generic resolves to `true` when `TF` is a digit that is less than `UF`. E.g.
> 
> * `TF` is `3` and `UF` is `5` -> `true`
> * `TF` is `3` and `UF` is `3` -> `false`
> * `TF` is `3` and `UF` is `1` -> `false`
> 
> Let's check how it works by example:
> 
> * when `TF` is `3` and `UF` is `5` it checks if `...3...5...` matches `0123456789`. It **is**, so `true`
> * when `TF` is `3` and `UF` is `3` it checks if `...3...3...` matches `0123456789`. It is **not**, so `false`
> * when `TF` is `3` and `UF` is `1` it checks if `...3...1...` matches `0123456789`. It is **not**, so `false`
> 
> I hope the explanation is sufficient :) Please ask if it not

This is super clear!! Thank you so much @Alexsey  :  )

**Comment by Barrenboat** [comment]
```ts
"0123456789" extends `${string}${TF}${string}${UF}${string}`
  ? false
  : true
```
genius code !
